kafka/client_async.py

```diff
@@ -588,6 +588,9 @@ def poll(self, timeout_ms=None, future=None):
                         metadata_timeout_ms,
                         idle_connection_timeout_ms,
                         self.config['request_timeout_ms'])
+                    # if there are no requests in flight, do not block longer than the retry backoff
+                    if self.in_flight_request_count() == 0:
+                        timeout = min(timeout, self.config['retry_backoff_ms'])
                     timeout = max(0, timeout / 1000)  # avoid negative timeouts
```

Comment:
It feels a little weird that this conflates both avoiding negative timeouts and changing the units from ms to seconds. The code works fine as-is, it just required me to re-read it twice to understand what was happening, especially given the inline comment. I would probably delay the conversion into seconds until `self._poll(timeout / 1000)` here: https://github.com/dpkp/kafka-python/pull/1823/files#diff-f161ac626efaecd3719ea5d72129604cR596 That way you keep it consistent as milliseconds until the last possible moment.

Reply:
makes sense -- worth doing that in a separate PR

Reply:
Done in https://github.com/dpkp/kafka-python/pull/1825, will merge once tests pass (to make sure I'm not overlooking something)